logstash-core/src/main/java/org/logstash/config/ir/compiler/DatasetCompiler.java

```diff
@@ -128,15 +134,58 @@ private DatasetCompiler() {
     }
 
     /**
-     * <p>Builds a terminal {@link Dataset} from the given parent {@link Dataset}s.</p>
+     * <p>Builds a terminal {@link Dataset} for the filters from the given parent {@link Dataset}s.</p>
+     * <p>If the given set of parent {@link Dataset} is empty the sum is defined as the
+     * trivial dataset that does not invoke any computation whatsoever.</p>
+     * {@link Dataset#compute(RubyArray, boolean, boolean)} is always
+     * {@link Collections#emptyList()}.
+     * @param parents Parent {@link Dataset} to sum
+     * @return Dataset representing the sum of given parent {@link Dataset}
+     */
+    public static Dataset terminalFilterDataset(final Collection<Dataset> parents) {
+        final int count = parents.size();
+        if (count > 1) {
+            final ClassFields fields = new ClassFields();
+            final ValueSyntaxElement outputBuffer = fields.add(new ArrayList<>());
+            final Closure clear = Closure.wrap();
+            final Closure compute;
+            final Collection<ValueSyntaxElement> parentFields = parents
+                .stream()
+                .map(fields::add)
+                .collect(Collectors.toList());
+            @SuppressWarnings("rawtypes") final RubyArray inputBuffer = RubyUtil.RUBY.newArray();
+            clear.add(clearSyntax(parentFields));
+            final ValueSyntaxElement inputBufferField = fields.add(inputBuffer);
+            compute = withInputBuffering(
+                Closure.wrap(
+                    // pass thru filter
+                    buffer(outputBuffer, inputBufferField)
+                ),
+                parentFields,
+                inputBufferField
+            );
+            return prepare(withOutputBuffering(compute, clear, outputBuffer, fields)).instantiate();
+        } else if (count == 1) {
+            // No need for a terminal dataset here, if there is only a single parent node we can
+            // call it directly.
+            return parents.iterator().next();
+        } else {
```

Comment:
personal taste: if this is like an "assert" I would probably put as first line, like:
```java
if (parents.isEmpty()) {
    throw new IllegalArgumentException(
           "Cannot create terminal filter dataset for an empty number of parent datasets"
    );
}
```

Reply:
Agree. will also do the same for both `terminalFilterDataset`  and `terminalOutputDataset`